marker_adapter/src/context.rs

```diff
@@ -54,92 +135,72 @@ impl<'ast> DriverContextWrapper<'ast> {
+    fn span(&'ast self, owner: SpanId) -> &'ast Span<'ast>;
+    fn span_snippet(&'ast self, span: &Span<'_>) -> Option<&'ast str>;
+    fn span_source(&'ast self, span: &Span<'_>) -> SpanSource<'ast>;
+    fn span_expn_info(&'ast self, expn_id: ExpnId) -> Option<&'ast ExpnInfo<'ast>>;
+    fn span_pos_to_file_loc(&'ast self, file: &FileInfo<'ast>, pos: SpanPos) -> Option<FilePos<'ast>>;
+    fn symbol_str(&'ast self, api_id: SymbolId) -> &'ast str;
+    fn resolve_method_target(&'ast self, id: ExprId) -> ItemId;
 }
 
-extern "C" fn body<'ast>(data: &'ast (), id: BodyId) -> &'ast Body<'ast> {
-    unsafe { as_driver_cx(data) }.body(id)
+extern "C" fn emit_diag<'a, 'ast>(data: &'ast (), diag: &Diagnostic<'a, 'ast>) {
+    unsafe { as_driver(data) }.emit_diag(diag);
 }
 
 extern "C" fn resolve_ty_ids<'ast>(data: &'ast (), path: ffi::FfiStr<'_>) -> ffi::FfiSlice<'ast, TyDefId> {
-    unsafe { as_driver_cx(data) }.resolve_ty_ids((&path).into()).into()
+    unsafe { as_driver(data) }.resolve_ty_ids((&path).into()).into()
 }
 
 // False positive because `SemTyKind` is non-exhaustive
 #[allow(improper_ctypes_definitions)]
 extern "C" fn expr_ty<'ast>(data: &'ast (), expr: ExprId) -> SemTyKind<'ast> {
-    unsafe { as_driver_cx(data) }.expr_ty(expr)
+    unsafe { as_driver(data) }.expr_ty(expr)
 }
 
 extern "C" fn span<'ast>(data: &'ast (), span_id: SpanId) -> &'ast Span<'ast> {
-    unsafe { as_driver_cx(data) }.span(span_id)
+    unsafe { as_driver(data) }.span(span_id)
 }
 
 extern "C" fn span_snippet<'ast>(data: &'ast (), span: &Span<'ast>) -> ffi::FfiOption<ffi::FfiStr<'ast>> {
-    unsafe { as_driver_cx(data) }.span_snippet(span).map(Into::into).into()
+    unsafe { as_driver(data) }.span_snippet(span).map(Into::into).into()
 }
 
 // False positive because `SpanSource` is non-exhaustive
 #[allow(improper_ctypes_definitions)]
 extern "C" fn span_source<'ast>(data: &'ast (), span: &Span<'_>) -> SpanSource<'ast> {
-    unsafe { as_driver_cx(data) }.span_source(span)
+    unsafe { as_driver(data) }.span_source(span)
 }
 
 extern "C" fn span_pos_to_file_loc<'ast>(
     data: &'ast (),
     file: &FileInfo<'ast>,
     pos: SpanPos,
 ) -> ffi::FfiOption<FilePos<'ast>> {
-    unsafe { as_driver_cx(data) }.span_pos_to_file_loc(file, pos).into()
+    unsafe { as_driver(data) }.span_pos_to_file_loc(file, pos).into()
 }
 
 extern "C" fn span_expn_info<'ast>(data: &'ast (), expn_id: ExpnId) -> FfiOption<&'ast ExpnInfo<'ast>> {
-    unsafe { as_driver_cx(data) }.span_expn_info(expn_id).into()
+    unsafe { as_driver(data) }.span_expn_info(expn_id).into()
 }
 
 extern "C" fn symbol_str<'ast>(data: &'ast (), sym: SymbolId) -> ffi::FfiStr<'ast> {
-    unsafe { as_driver_cx(data) }.symbol_str(sym).into()
+    unsafe { as_driver(data) }.symbol_str(sym).into()
 }
 
 extern "C" fn resolve_method_target<'ast>(data: &'ast (), id: ExprId) -> ItemId {
-    unsafe { as_driver_cx(data) }.resolve_method_target(id)
+    unsafe { as_driver(data) }.resolve_method_target(id)
```

Comment:
Technically `unsafe` must be propagated to the surrounding `extern "C"` functions, because they bear an implicit contract that `data` is a wellformed pointer to a `MarkerContextWrapper`. If the functions are not marked as `unsafe` it means you can call the `resolve_method_target`, for example like this from safe Rust and trigger undefined behavior:
```rust
resolve_method_target(&(), some_expr_id)
```
So the unsafe invariant must be propagated to the caller up to that point where safe Rust knows for sure it passes a wellformed pointer to `MarkerContextWrapper` and no other invoking safe Rust code can break that.

Reply:
You might be right on this one. Thinking about it, it would be better to have a dummy struct inside `marker_api`, like `MarkerContextDriverData` and use that for the reference instead. So that would make it `&'ast MarkerContextDriverData` instead of `&'ast ()`. I like this more than the current `&()` one, and it's harder to mess up. In that case, the function wouldn't need to be unsafe either, right :thinking: